File: 8-string-to-integer-atoi/string-to-integer-atoi.cpp
class Solution {
public:
    int myAtoi(string s) {
        bool flag=false;
        bool neg=false;
        int res = 0;
        int i = 0;
        while(s[i] != '\0'){
            if(!flag){
                if(s[i] == ' '){    i++;     continue;}
                if(s[i] == '+'){     neg = false;   flag=true; }
                else if(s[i] == '-'){     flag=true;     neg=true;   }
                else if(s[i] >= 48 && s[i] <= 57){
                    flag = true;    
                    res = int(s[i])-48;
                }   
                else    return res;
            }
            else{
                if(s[i] == '0' && res == 0){    i++;    continue; }
                else if(s[i] >= 48 && s[i] <= 57){
                    if(!neg && ((INT_MAX/10) == res) && (int(s[i])-48) < 7 && s[i+1] == '\0')     return 10*res+(int(s[i])-48);    
                    else if(!neg && ((INT_MAX/10)-res) <= (int(s[i])-48))  return INT_MAX;
                    else if(neg && (INT_MAX/10 == res) && (int(s[i])-48)<8 && s[i+1] == '\0') return  -1*(10*res+(int(s[i])-48));
                    else if(neg && ((INT_MAX/10-res) <= (int(s[i])-47))){    return INT_MIN; }
                    else{    res = res*10+int(s[i])-48; }
                }
                else  break;
            }
            i++;
        }
        if(neg) return -1*res;
        return res;
    }
};

File: 8-string-to-integer-atoi/string-to-integer-atoi.py
class Solution:
    digits = 0
    def Atoi(self, s:str, sign:int):
        if(len(s) == 0):
            return 0
        # if(len(s) == 1 and )
        
        if(s[0] in [' ', '+', '-']):
            return 0
        
        if(s[0] >= '0' and s[0] <= '9'):
            res = self.Atoi(s[1:], sign)
            self.digits += 1
            if(sign == -1):
                return min(10**(self.digits-1)*int(s[0])+res, 2**(31))
            else:
                return min(10**(self.digits-1)*int(s[0])+res, 2**(31)-1)
        return 0
    
    def myAtoi(self, s: str) -> int:
        if(len(s) == 0):
            return 0
        while(s[0] == ' '):
            return self.myAtoi(s[1:])
        if(s[0] == '-'):
            sign = -1
            return -1*self.Atoi(s[1:], sign)
        if(s[0]== '+'):
            sign = 1
            return self.Atoi(s[1:], sign)
        return self.Atoi(s, 1)
